File: SRM 657/ProblemSets.java
import java.io.*;
import java.util.*;

public class ProblemSets {
    private boolean f(long e, long em, long m, long mh, long h, long gap) {
        if (gap > e + em) return false;
        if (gap > h + mh) return false;
        em -= Math.max((long) 0, gap - e);
        mh -= Math.max((long) 0, gap - h);
        return (m + em + mh) >= gap;

    }
	public long maxSets(long E, long EM, long M, long MH, long H) {
        long l = 0, r = Long.MAX_VALUE;
        while (l + 1 < r) {
            long m = l + (r - l) / 2;
            if (f(E, EM, M, MH, H, m)) {
                l = m;
            } else {
                r = m;
            }
        }
		return l;
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("ProblemSets (250 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            long E = Long.parseLong(Reader.nextLine());
            long EM = Long.parseLong(Reader.nextLine());
            long M = Long.parseLong(Reader.nextLine());
            long MH = Long.parseLong(Reader.nextLine());
            long H = Long.parseLong(Reader.nextLine());
            Reader.nextLine();
            long __answer = Long.parseLong(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(E, EM, M, MH, H, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1500341561;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 250 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(long E, long EM, long M, long MH, long H, long __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		ProblemSets instance = new ProblemSets();
		long __result = 0;
		try {
			__result = instance.maxSets(E, EM, M, MH, H);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "ProblemSets.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: SRM 657/ProblemSets.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
typedef long long ll;
bool f(ll a, ll b, ll c, ll d, ll e, ll gap) {
    if (gap > a) {
        ll tmp = gap - a;
        if (b < tmp) return false;
        b -= tmp;
    }
    if (gap > e) {
        ll tmp = gap - e;
        if (d < tmp) return false;
        d -= tmp;
    }
    c += (b + d);
    return c >= gap;
}
class ProblemSets {
    public:
    long long maxSets(long long E, long long EM, long long M, long long MH, long long H) {
        ll l = 0, r = 0x7f7f7f7f7f7f7f7f;
        while (l + 1 < r) {
            ll m = (l + r) / 2;
            if (f(E,EM,M,MH,H,m)) {
                l = m;
            } else {
                r = m;
            }
        //    cout << "l=" << l << ",r=" << r << endl;
        }

        return l;
    }
};

// CUT begin
ifstream data("ProblemSets.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(long long E, long long EM, long long M, long long MH, long long H, long long __expected) {
    time_t startClock = clock();
    ProblemSets *instance = new ProblemSets();
    long long __result = instance->maxSets(E, EM, M, MH, H);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        long long E;
        from_stream(E);
        long long EM;
        from_stream(EM);
        long long M;
        from_stream(M);
        long long MH;
        from_stream(MH);
        long long H;
        from_stream(H);
        next_line();
        long long __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(E, EM, M, MH, H, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1500341897;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 250 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "ProblemSets (250 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
